core/src/engine/single_waiter_mutex.cpp

```diff
@@ -0,0 +1,35 @@
+#include <userver/engine/single_waiter_mutex.hpp>
+
+#include <engine/impl/mutex_impl.hpp>
+
+USERVER_NAMESPACE_BEGIN
+
+namespace engine {
+
+class SingleWaiterMutex::ImplWrapper final {
+public:
+  using Impl = impl::MutexImpl<impl::WaitListLight>;
+
+  Impl& operator->() {return impl;};
```

Comment:
`operator->` should return a pointer. Not sure if that wrapper is much better than just empty class publicly inherited from `MutexImpl` though.

Reply:
The main disadvantage of the approach with an inheritance: not final MutexImpl without virtual destructor is a potential risk. MutexImpl with virtual destructor requires additional memory. The wrapper approach is safe and cost-free.

Reply:
As this is not a part of public interface, non-public inheritance is okay as long as we don't store these as base class pointers.

It is a matter of taste here though.